src/main/java/com/google/firebase/ThreadManager.java

```diff
@@ -36,42 +36,71 @@
 public abstract class ThreadManager {
 
   @NonNull
-  final ListeningScheduledExecutorService getListeningExecutor(@NonNull FirebaseApp app) {
-    ScheduledExecutorService executor = getExecutor(app);
-    checkNotNull(executor, "ScheduledExecutorService must not be null");
-    return MoreExecutors.listeningDecorator(executor);
+  final FirebaseExecutor getFirebaseExecutor(@NonNull FirebaseApp app) {
+    return new FirebaseExecutor(getExecutor(app));
+  }
+
+  final void releaseFirebaseExecutor(@NonNull FirebaseApp app, @NonNull FirebaseExecutor executor) {
+    releaseExecutor(app, executor.delegate);
   }
 
   /**
    * Returns the main thread pool for an app. Implementations may return the same instance of
-   * <code>ScheduledExecutorService</code> for multiple apps. The returned thread pool is used by
-   * all components of an app except for the Realtime Database. Database has far stricter and
-   * complicated threading requirements, and thus initializes its own threads using the
-   * factory returned by {@link ThreadManager#getThreadFactory()}.
+   * <code>ExecutorService</code> for multiple apps. The returned thread pool is used for
+   * short-lived tasks by all components of an app.
+   *
+   * <p>For long-lived tasks (such as the ones started by the Realtime Database client), the SDK
+   * creates dedicated executors using the <code>ThreadFactory</code> returned by the
+   * {@link #getThreadFactory()} method.
    *
    * @param app A {@link FirebaseApp} instance.
-   * @return A non-null {@link ScheduledExecutorService} instance.
+   * @return A non-null {@link ExecutorService} instance.
    */
   @NonNull
-  protected abstract ScheduledExecutorService getExecutor(@NonNull FirebaseApp app);
+  protected abstract ExecutorService getExecutor(@NonNull FirebaseApp app);
```

Comment:
It's unclear to me when what the difference between and Executor and a FirebaseExecutor is.

Reply:
`FirebaseExecutor` is just a holder for the `ListeningExecutorService` **and** the underlying `ExecutorService` returned by the user code. By keeping a reference to the original `ExecutorService` around we can make sure that the argument passed to `releaseExecutor()` is same as the one returned by `getExecutor()`.

Reply:
I don't know if we want to go down that road - but what if we required the user to provide a ListeningExecutorService in the first place? Would that be cleaner?

Reply:
It would make things simpler for us. But we would be putting the onus of managing these references on the user. I would also prefer not to tie our API to a Guava type.